aten/src/ATen/native/TensorShape.cpp

```diff
@@ -392,6 +390,94 @@ Tensor cat(TensorList tensors, int64_t dim) {
   return result;
 }
 
+Tensor block_diag(TensorList tensors) {
+  Tensor result;
+  if (tensors.size() == 0) {
+    result = at::empty({1, 0});
+    return result;
+  }
+
+  // Use at::native::result_type to find the right tensor to copy options from, so
+  // that the output data type is correct
+  const Tensor* options_tensor = &tensors[0];
+  ScalarType output_scalar_type = options_tensor->scalar_type();
+  const Device& device = tensors[0].device();
+
+  for (size_t tensor_idx = 1; tensor_idx < tensors.size(); tensor_idx++) {
+    const Tensor* other_tensor = &tensors[tensor_idx];
+
+    TORCH_CHECK(
+      other_tensor->device() == device,
+      "torch.block_diag: input tensors must all be on the same device.",
+      " Input 0 is on device ", device,
+      " and input ", tensor_idx, " is on device ", other_tensor->device()
+    );
+
+    ScalarType scalar_type = at::native::result_type(*options_tensor, *other_tensor);
```

Comment:
Sorry, when I asked if we could use `native::result_type`, I was pointing to the `native::result_type(TensorList)` overload. So instead of calling result_type on two tensors for each tensor in the loop, we could straight up compute the result type in one go:
```
// outside of the for loop
output_scalar_type = native::result_type(tensors).
```
That would save us from having to reimplement the logic of finding the result type of all of the tensors.

It doesn't actually save too much, so I'm also fine if you want to keep this as-is

Reply:
Oh I didn't realize that's what you meant. That is a better way to go.